addons/purchase/models/purchase_order.py

```diff
@@ -152,8 +152,11 @@ def _compute_invoice(self):
     @api.constrains('company_id', 'order_line')
     def _check_order_line_company_id(self):
         for order in self:
-            product_company = order.order_line.product_id.company_id
-            companies = product_company and product_company._accessible_branches()
+            product_companies = order.order_line.product_id.company_id
+            companies = self.env['res.company']
+            if product_companies:
+                for product_company in product_companies:
+                    companies |= product_company._accessible_branches()
```

Comment:
It seems to me that this is not checking what we want anymore.
If the order company is in the accessible branches of any product company, it will be considered valid.  So if there is only one accessible product in the order, it won't raise.
But the order company should be in the accessible branches of each of the product companies.